service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/AgentDirectoryMapper.java

```diff
@@ -0,0 +1,140 @@
+            .observationOrganizations(prepareObservationAgentsList(bundle))
+            .referralRequestsOrganizations(prepareReferralRequestAgentsList(bundle))
+            .agentPersons(preparePractitionerRoleAgentsList(bundle));
+
+        return TemplateUtils.fillTemplate(AGENT_DIRECTORY_STRUCTURE_TEMPLATE, builder.build());
+    }
+
+    private String extractPatientManagingOrganization(Bundle bundle, String nhsNumber) {
+        return AgentDirectoryExtractor.extractPatientByNhsNumber(bundle, nhsNumber)
+            .filter(Patient::hasManagingOrganization)
+            .map(patient -> ResourceExtractor.extractResourceByReference(bundle,
+                patient.getManagingOrganization().getReferenceElement()))
+            .map(Optional::get)
+            .map(Organization.class::cast)
+            .map(organizationToAgentMapper::mapOrganizationToAgent)
+            .orElseThrow(() -> new EhrMapperException("No patient or managing organization found in ehrFolder with NHS Number: "
+                + nhsNumber));
+    }
+
+    private List<String> prepareObservationAgentsList(Bundle bundle) {
+        List<Observation> observations = AgentDirectoryExtractor.extractObservationsWithPerformers(bundle);
+
+        return observations.stream()
+            .map(observation -> buildAgentPerson(observation, bundle))
+            .filter(StringUtils::isNotEmpty)
+            .collect(Collectors.toList());
+    }
+
+    private List<String> prepareReferralRequestAgentsList(Bundle bundle) {
+        List<ReferralRequest> referralRequests = AgentDirectoryExtractor.extractReferralRequestsWithRequester(bundle);
+
+        return referralRequests.stream()
+            .map(referralRequest -> buildAgentPerson(referralRequest, bundle))
+            .filter(StringUtils::isNotEmpty)
+            .collect(Collectors.toList());
+    }
+
+    private List<String> preparePractitionerRoleAgentsList(Bundle bundle) {
+        List<Practitioner> practitioners = AgentDirectoryExtractor.extractRemainingPractitioners(bundle);
+        var mappingTriplets = AgentDirectoryExtractor.extractPractitionerRoleTriples(bundle, practitioners);
+
+        return mappingTriplets.stream()
+            .map(triple -> mapAgentPerson(triple.getLeft(), triple.getMiddle(), triple.getRight()))
+            .filter(StringUtils::isNotEmpty)
+            .collect(Collectors.toList());
+    }
+
+    private String buildAgentPerson(Observation observation, Bundle bundle) {
+        var practitioner = AgentDirectoryExtractor.extractObservationResource(observation, bundle, ResourceType.Practitioner);
+        var organization = AgentDirectoryExtractor.extractObservationResource(observation, bundle, ResourceType.Organization);
+        if (practitioner.isPresent() && organization.isPresent()) {
+            return mapAgentPerson((Practitioner) practitioner.get(),
+                Optional.empty(),
+                organization.map(value -> (Organization) value));
```

Comment:
If you like the `::cast` thing
```suggestion
            return mapAgentPerson(practitioner.map(Practitioner.class::cast).get(),
                Optional.empty(),
                organization.map(Organization.class::cast));
```